### PR TITLE
upload global scatter op files

### PR DESCRIPTION
**Pull request by youth123**
### PR types

New features
### PR changes

OPs
### Describe

Add global scatter operator to help implement moe function.

### CODE REVIEW

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+namespace paddle {
+namespace operators {
+
+class GlobalScatterOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
+    OP_INOUT_CHECK(ctx->HasInput("x"), "Input", "x", "GlobalScatter");
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: Capital X？We usually use X in C++ side op.

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                   "GlobalScatter");
+    OP_INOUT_CHECK(ctx->HasOutput("Out"), "Output", "Out", "GlobalScatter");
+    int ring_id = ctx->Attrs().Get<int>("ring_id");
+    PADDLE_ENFORCE_GE(
+        ring_id, 0,
+        platform::errors::InvalidArgument(
+            "The ring_id (%d) for global scatter op must be non-negative.",
+            ring_id));
+    // int in_feat = ctx->Attrs().Get<int>("in_feat");
+    // VLOG(0) << "scatter in_feat " << in_feat;
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: Delete comments.

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+      const framework::ExecutionContext& ctx) const override {
+    return framework::OpKernelType(
+        OperatorWithKernel::IndicateVarDataType(ctx, "x"), ctx.GetPlace());
+  }
+};
+
+class GlobalScatterOpMaker : public framework::OpProtoAndCheckerMaker {
+ public:
+  void Make() {
+    AddInput("x", "(Tensor) tensor send.");
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: Capital X.

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    AddInput("local_count",
+             "(Tensor) Tensor which have n_expert * world_size elements that "
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: have --> has

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+             "indicates"
+             "how many data needed to be sent.");
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: how many data needed to be sent to each expert

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    AddInput("global_count",
+             "(Tensor) Tensor which have n_expert * world_size elements that "
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: have --> has

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+             "indicates"
+             "how many data needed to be received.");
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: how many data needed to receive from each expert

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    // AddAttr<int>("in_feat", "(int default 0) The feature shape of the x.")
+    //     .SetDefault(0);
+    // AddAttr<int>("n_expert", "(int default 1) The number of experts in every
+    // rank.")
+    //     .SetDefault(1);
+    // AddAttr<int>("world_size", "(int default 1) The number of rank.")
+    //     .SetDefault(1);
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: Delete the above comments

File: paddle/fluid/operators/collective/global_scatter_op.cc [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+ protected:
+  void Apply(GradOpPtr<T> retv) const override {
+    retv->SetType("global_gather");
+    retv->SetInput("x", this->OutputGrad("Out"));
+    retv->SetInput("local_count", this->Input("local_count"));
+    retv->SetInput("global_count", this->Input("global_count"));
+    retv->SetOutput("Out", this->InputGrad("x"));
+    retv->SetAttrMap(this->Attrs());
+    // retv->SetAttr("in_feat", retv->GetAttr("in_feat"));
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cc]: Delete comments

File: paddle/fluid/operators/collective/global_scatter_op.cu.cc [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+namespace paddle {
+namespace operators {
+template <typename T>
+class GlobalScatterOpCUDAKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+#if defined(PADDLE_WITH_NCCL)
+#if NCCL_VERSION_CODE >= 2703
+    VLOG(0) << "scatter start";
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cu.cc]: Use VLOG(x) where x > 0, VLOG(3) is recommended to use.

File: paddle/fluid/operators/collective/global_scatter_op.cu.cc [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    auto x = ctx.Input<framework::LoDTensor>("x");
+    auto local_count = ctx.Input<framework::LoDTensor>("local_count");
+    auto global_count = ctx.Input<framework::LoDTensor>("global_count");
+    // auto in_feat = static_cast<int>(ctx.Attr<int>("in_feat"));
+    // auto n_expert = static_cast<int>(ctx.Attr<int>("n_expert"));
+    // auto world_size = static_cast<int>(ctx.Attr<int>("world_size"));
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cu.cc]: Delete above comments.

File: paddle/fluid/operators/collective/global_scatter_op.cu.cc [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    auto out = ctx.Output<framework::LoDTensor>("Out");
+
+    framework::Tensor cpu_local_count;
+    framework::TensorCopy(*local_count, platform::CPUPlace(), &cpu_local_count);
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cu.cc]: 是不是可以加个判断，如果local_count本身就在cpu上，则不做拷贝。

File: paddle/fluid/operators/collective/global_scatter_op.cu.cc [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    auto comm = platform::NCCLCommContext::Instance().Get(ring_id, place);
+    cudaStream_t stream = nullptr;
+    if (ctx.Attr<bool>("use_calc_stream")) {
+      auto dev_ctx = platform::DeviceContextPool::Instance().Get(place);
+      stream = static_cast<platform::CUDADeviceContext*>(dev_ctx)->stream();
+    } else {
+      stream = comm->stream();
+    }
+    int nranks = comm->nranks();
+    auto in_feat = x->dims()[1];
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cu.cc:75]: 这个地方的假设是x为二维的，需要加个合法性校验。

File: paddle/fluid/operators/collective/global_scatter_op.cu.cc [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    auto n_expert = local_count->dims()[0] / nranks;
+    auto fwd_count = 0;
+    auto global_count_len = cpu_global_count.numel();
+    for (auto i = 0; i < global_count_len; ++i)
+      fwd_count += cpu_global_count_data[i];
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cu.cc:80]: 一般建议for的语句块用{}包起来。

File: paddle/fluid/operators/collective/global_scatter_op.cu.cc [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+          PADDLE_ENFORCE_CUDA_SUCCESS(
+              platform::dynload::ncclRecv(recv_buf + recv_ptr * in_feat,
+                                          cpu_global_count_data[idx] * in_feat,
+                                          dtype, j, comm->comm(), stream));
+          recv_ptr += cpu_global_count_data[idx];
+        }
+      }
+      PADDLE_ENFORCE_CUDA_SUCCESS(platform::dynload::ncclGroupEnd());
+      PADDLE_ENFORCE_CUDA_SUCCESS(cudaStreamSynchronize(stream));
+      VLOG(0) << "scatter end";
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.cu.cc]: VLOG(3)

File: paddle/fluid/operators/collective/global_scatter_op.h [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#include "paddle/fluid/framework/data_type.h"
+#include "paddle/fluid/framework/lod_tensor.h"
+#include "paddle/fluid/framework/op_registry.h"
+
+#if defined(PADDLE_WITH_GLOO)
+#include "paddle/fluid/framework/fleet/gloo_wrapper.h"
```
Comment (sandyhouse) [inline paddle/fluid/operators/collective/global_scatter_op.h:21]: 这三行不需要。

File: python/paddle/distributed/utils.py [hunk 17]
```diff
@@ -42,9 +66,111 @@
...
-           'pull_worker_log'
+           'pull_worker_log',
+           'global_scatter'
 ]
 
+
+def global_scatter(x, \
+                   local_count, \
+                   global_count, \
+                   group=None, \
```
Comment (sandyhouse) [inline python/paddle/distributed/utils.py]: 后面的'\'符号不需要

Comment (youth123) [reply python/paddle/distributed/utils.py]: 是不需要#noqa注释嘛

Comment (sandyhouse) [reply python/paddle/distributed/utils.py]: 后面的反斜线符号不需要。

File: paddle/fluid/operators/collective/global_scatter_op.cu.cc (hunk 13 above)
Comment (youth123) [reply paddle/fluid/operators/collective/global_scatter_op.cu.cc:75]: 在infer shape函数里加了对输入维度判断。

**Pull request closed without merging**